expr/shaper.go

```diff
@@ -217,154 +217,163 @@ func (s *step) buildRecord(in zcode.Bytes, b *zcode.Builder) {
-	typ        zng.Type
-	shapeSpecs map[int]shapeSpec // map from type ID to shapeSpec
+	expr       Evaluator
+	shapeTo    zng.Type
+	shapers    map[int]*shaper // map from input type ID to shaper
 	transforms ShaperTransform
 }
 
-// NewShaperType returns a shaper that will shape the result of fieldExpr
-// to the provided typExpr. (typExpr should evaluate to a type value,
-// e.g. a value of type TypeType).
-func NewShaper(zctx *resolver.Context, fieldExpr Evaluator, typ zng.Type, tf ShaperTransform) (*Shaper, error) {
-	return &Shaper{
+// NewConstShaper returns a shaper that will shape the result of expr
+// to the provided shapeTo type.
+func NewConstShaper(zctx *zson.Context, expr Evaluator, shapeTo zng.Type, tf ShaperTransform) *ConstShaper {
+	return &ConstShaper{
 		zctx:       zctx,
-		fieldExpr:  fieldExpr,
-		typ:        typ,
-		shapeSpecs: make(map[int]shapeSpec),
+		expr:       expr,
+		shapeTo:    shapeTo,
+		shapers:    make(map[int]*shaper),
 		transforms: tf,
-	}, nil
+	}
 }
 
-// NewShaperFromTypeExpr returns a shaper that will shape the result of fieldExpr
-// to the provided typExpr. (typExpr should evaluate to a type value,
-// e.g. a value of type TypeType).
-func NewShaperFromTypeExpr(zctx *resolver.Context, fieldExpr, typExpr Evaluator, tf ShaperTransform) (*Shaper, error) {
-	switch typExpr.(type) {
-	case *Var, *TypeFunc, *Literal:
-	default:
-		return nil, fmt.Errorf("shaping functions (crop, fill, cast, order) require a type value as second parameter")
+type Shaper struct {
+	zctx       *zson.Context
+	typExpr    Evaluator
+	expr       Evaluator
+	shapers    map[zng.Type]*ConstShaper
+	transforms ShaperTransform
+}
+
+// NewShaper returns a shaper that will shape the result of expr
+// to the type returned by typExpr.
+func NewShaper(zctx *zson.Context, expr, typExpr Evaluator, tf ShaperTransform) *Shaper {
+	return &Shaper{
+		zctx:       zctx,
+		typExpr:    typExpr,
+		expr:       expr,
+		shapers:    make(map[zng.Type]*ConstShaper),
+		transforms: tf,
 	}
-	// XXX This needs to be done at runtime because the type expression can
-	// depend on typenames defined in the data.  See issue #2383.
-	typVal, err := typExpr.Eval(nil)
+}
+
+func (s *Shaper) Eval(rec *zng.Record) (zng.Value, error) {
+	typVal, err := s.typExpr.Eval(rec)
 	if err != nil {
-		return nil, err
+		return zng.Value{}, err
 	}
 	if typVal.Type != zng.TypeType {
-		return nil, fmt.Errorf("shaping functions (crop, fill, cast, order) require a type value as second parameter")
-	}
-	s, err := zng.DecodeString(typVal.Bytes)
-	if err != nil {
-		return nil, err
+		return zng.NewErrorf("shaper function type argument is not a type"), nil
 	}
-	shapeType, err := zctx.Context.LookupByName(s)
-	if err != nil {
-		return nil, fmt.Errorf("shaper could not parse type value literal: %s", err)
+	shapeTo := s.zctx.FromTypeValue(typVal)
+	if shapeTo == nil {
+		return zng.NewErrorf("shaper encountered unknown type value: %q", string(typVal.Bytes)), nil
```

Comment:
Yes, malformed.  The zcode.Bytes should be a ZSON string so I wanted to just print the string.  Without the cast, the zcode.Bytes stringer method will be called and it will print the zcode byte string.

Reply:
I forgot about zcode.Bytes.String!

Reply:
The change below makes this go away anyway...